### ISSUE TITLE
[Discussion] - [Custom hyperopt loss plug and play rules]

### ISSUE THREAD

**Issue by gaugau3000** [freqtrade/freqtrade#2270]
As a trader
I want to pick some existing rules to optimize my strategy and give them a weight
So that i gain time when creating my own custom hyperopt loss file.

Below an exemple on combination of calmar ratio, nb expected trade = 600, r_square of a linear regression (prevent overfitting with peaks in percent_results) with different weight.

![image](https://user-images.githubusercontent.com/2643583/65068918-e6f3e000-d989-11e9-8eed-2549cb2e1c07.png)

![image](https://user-images.githubusercontent.com/2643583/65069068-35a17a00-d98a-11e9-82a7-6616b2ea8a1f.png)

Here i use an arbitrary (-0.5,0.5) centered in 0 to mix rules together.

Have a nice day.

**Comment by xmatthias** [comment]
While this seems promising from a mathematical standpoind, i'm not certain sure how you would implement this in practice, since realistically, you'd need a hyperopt loss function for every possible combination. 

In your example it's 3 parameters - but maybe i don't want one parameter, so we should account for this - or i want another parameter not listed here.

I currently don't see how this can be generalized to work for different loss function types (where i can use Calmar or Sharpe or ...) in combination with any of the other parameters, unless Calmar (and Sharpe, and others...) does implement these custom options themselfs as proposed as example in https://github.com/freqtrade/freqtrade/pull/2255#issuecomment-530951378 and https://github.com/freqtrade/freqtrade/pull/2255#issuecomment-530933345.

Some helper-functions can be added to calculate some "default" metrics if desired - but if SharpeLoss does not implement these, then configuring these wont have an active effect.

Maybe you have an architecture on how you'd like to implement this in mind ?

**Comment by gaugau3000** [comment]
We can provide functions to the end user that he can use in is custom hyperopt loss class. If he want to optimize using calmar and an expected number of trade he just need to do something like that.

 **End user**

_hyperopt_loss_custom.py_
```

"""
Normalize loss value to be float between (-0.5, 0.5)
0 mean no profit
"""
calmar_loss = compute_norm_simulated_calmar_ratio(results, weight=CALMAR_WEIGHT)

"""
Normalize loss value to be float between (0, 0.5) :
0 mean trade_count = NUM_TRADE_WEIGHT
"""
expected_trade_penalty_loss = compute_norm_expected_trade_loss_penalty(
results,
weight=NUM_TRADE_WEIGHT, 
trade_expected_number=NUM_TRADES)

return calmar_loss + expected_trade_penalty_loss 

```
It can interface with current work of @hroff-1902 
--hyperopt-loss-params="{'CALMAR_WEIGHT'=0.5,'NUM_TRADE_WEIGHT'=0.5, 'NUM_TRADES'=200}"

Depending the weight it modify the curve slope but the value stays between (-0.5,0.5) so we can sum up the constraints 

**Developper**

- compute the calmar ratio. 
- normalize the calmar ratio (to ensure that he can be easily combine with other rules).

For the sharpeLoss instead of providing the loss class just provide a compute_norm_sharpe_ratio (based on a compute sharpe ratio) -> if the end user would like to test an optimization based on sharpe ratio he just do

_hyperopt_loss_custom.py_
```

return compute_norm_sharpe_ratio(results,weight=1)

```

in his custom loss class

For testing part it may be easier to test has the helpers function have a unique responsibility. I don't test the whole custom class that maybe contain many rules (except as integration/functional test).

Hope this help to have an overview.

Have a nice day.
